### ISSUE TITLE
Design calculate_scores_pipeline

### ISSUE THREAD

**Issue by florian-huber** [matchms/matchms#268]
In #251 we discussed how to add different similarity scores, or how to combine masking steps with similarity score computations.

One such option would be to introduce a function/class that allows users to easily stack multiple operations such that the computation time can be reduced significantly (when compared to a naive sequence of simliarity computations).
One example: Two large sets of spectra should be compared using a computationally expensive score (e.g. Cosine). But many pairs can easily be sorted our based on other criteria, e.g. their metadata. For instance: two spectra with very different precusors can be skipped etc.

For this we could implement something like this:
```python
from matchms import calculate_scores_pipeline, Spectrum
from matchms.similarity import CosineGreedy, MetadataMatch, PrecursorMzMatch

spectrums_1 = [...]
spectrums_2 = [...]

scores = calculate_scores_pipeline(spectrums_1, spectrums_2,
                                   masks=[MetadataMatch(field="retention_time", tolerance=60.0),
                                          PrecursorMzMatch(tolerance=50.0, tolerance_type="Dalton")],
                                   score=CosineGreedy())
```
This would first run `MetadataMatch()` on all-vs-all spectrums_1 vs spectrums_2 and fill a matrix with 1 (True) or 0 (False).  
For all True values of matrix: run `PrecursorMzMatch()` and update matrix.  
For all True values of matrix: run `CosineGreedy()`

**Comment by hechth** [comment]
I think it might also make sense to come back to #210 and separate the pair finding function (if we already split it into a pipeline). This way we could also easily implement a forward and reverse score.

**Comment by florian-huber** [comment]
Sorry, I am not sure I fully get the point.
The main issue I see for the pipeline described above is that we no longer need to compute the score for each possible pair. Only for those which pass the masking steps (say, all which have a retention time difference within a set tolerance).
For this I only need a function (or method) that can compute pair-wise scores. Those should already work for all `matchms.similariy` --> `similarity_score.pair(spectrum1, spectrum2)`.

Why would be need to go into the internals of the pair-wise computation?

**Comment by florian-huber** [comment]
Here a few more thoughts on possible designs.

**Option 1** stacking method for `Scores`
This is different from what was sketched above, but could work out as well.
```python
scores = calculate_scores(references=spectrums_1,
                          queries=spectrums_2,
                          similarity_functions=PrecursorMzMatch(tolerance=50.0, tolerance_type="Dalton"),
                          sparse=True)

scores.stack_scores(MetadataMatch(field="retention_time", tolerance=60.0), keep="last")
scores.stack_scores(CosineGreedy(), keep="last", threshold=0.5)

# that could change the Scores object:
print(scores.similarity_functions)  #=> [PrecursorMzMatch(), MetadataMatch(f), CosineGreedy()]
print(scores.thresholds)  # => [50.0, 60.0, 0.5]
```

`Scores.stack_scores()` would then be a method that takes the current `Scores.scores` either where `True` or above a given threshold and computes another similarity score for those pairs. Possible options could then be to say
- `keep="last"` to only keep the latest similarity score for those pairs (hence make the previous ones a mask)
- `keep="all"` to keep all scores for the pairs that satisfy the given criteria (e.g. stacked on top of each other)
- `keep="multiply"` or other alternatives could also be thought of, for instance to compute the product of all scores that fullfill the criteria etc. 


**Option 2** `calculate_scores_pipline()` function
This is what was outlined above.
```python
scores = calculate_scores_pipeline(spectrums_1, spectrums_2,
                                   masks=[MetadataMatch(field="retention_time", tolerance=60.0),
                                          PrecursorMzMatch(tolerance=50.0, tolerance_type="Dalton")],
                                   score=CosineGreedy())
```

---

Right now I tend towards option1. 
That would, however, require a few changes to `Scores`.
- attribute `similarity_function` must become `similarity_functions` which allows input of list of `BaseSimilarity` objects.
- additional attribute to store thresholds used for the different scores. Default can be set to None, which would give the results as they are now.

**Comment by florian-huber** [comment]
Further down the road I came across another issue.
The current matchms-compatible similarity score methods have to provide two methods: `.pair()` and `.matrix()`. For simple methods `.matrix()` will simply be a double-for loop over all pairs, but for more complicated+optimized measures than the actual implementations can differ a lot (e.g. `Spec2Vec`, `MS2DeepScore`).

If we now go into pipelines we might often just want to compute a fraction of all pairs. That could of course be implemented by looping over all pairs which are > threshold for another score (or mask). But from `matchms` we cannot decide if that is a good decision perfomance-wise!
(for some scores computing 10% of a matrix in a loop might take **way** longer than computing all scores with an optimized matric-method!)

So I would suggest we add a 3rd required method to `BaseSimilarity` which is called `.sparse_matrix()`. The default (= if not implemented separately) can fall back to `.matrix()`.
But developers are free to implement there own, more optimized versions.

**Comment by florian-huber** [comment]
I had another dive and now think that sparse matrices in `Scores` could be great for such a pipeline (#60). That would allow us to store several different scores on top of each other because they will not be kept as Numpy arrays, but as indices and values. And several scores for the same pairs would only mean several value arrays, but no more duplication of the indices.

This whole thing will potentially also be very usefull for tools such as [**MS2Query** ](https://github.com/iomega/ms2query/) so any feedback from that side is much appreciated @niekdejonge.

**Comment by niekdejonge** [comment]
Interesting plans! 

If I understand it correctly the main goals are to make creating new pipelines easier and cleaner and to speed up the score calculation?

One thing I wonder if, if it is doable to make an implementation in matchms that is flexible enough to be used for new tools, or that developers will in the end still develop there own methods, since they need functionalities that are slightly different. (see also suggested ideas needed for MS2Query below) I don't know the answer, but can imagine that might be challenging :) 

In MS2Query we de a preselection on MS2Deepscore followed by calculating 8 other scores. 
My current implementation of this method, is not by working with a sparse matrix but by just handling 1 query spectrum at a time and selecting the top 2000 spectra based on MS2Deepscore followed by calculating the scores between 1 query spectrum and the top 2000 spectra. 

If I understand it correctly the proposed matchms method would take in all query spectra at once, while making sure no scores are calculated that are not needed. My gut feeling says that this matchms method would be quicker, but I don't have a clear reason why and by how much. So what kind of speed optimization do you expect, doing it in this proposed way, compared to the way we have this currently implemented in MS2Query. (If it does not speed up the process, it would of course still make the code cleaner)

To make a pipeline function like this useful for MS2Query it would need the following functionalities:
- Instead of filtering on a threshold, make it also possible to select the top scoring spectra. 
- Make it possible to calculate more than one score and store in the same Scores class. (I just read back and think you already suggest this, in the comment above)
- Make it possible to easily create your own scoring function that can be used by the pipeline. (I think this is already implemented).

So for MS2Query this would result in running a function like:
calculate_scores_pipeline(all_query_spectra, all_reference_spectra,
                                   masks=[MetadataMatch(field="retention_time", top_scoring=2000)],
                                   scores=CosineGreedy(), Spec2Vec(), parent_mass_difference(), user_defined_method())

Resulting in a Scores object containing a partly filled matrix for all the provided scores. 

An implementation like this, would make developing new tools like MS2Query a lot cleaner and easier, but I am not really sure how easy it is to implement this in an efficient and flexible way in Matchms. 

Hope this is a bit helpful, let me know if something is unclear.

**Comment by florian-huber** [comment]
```python
scores = calculate_scores(references=spectrums_1,
                          queries=spectrums_2,
                          similarity_functions=PrecursorMzMatch(tolerance=50.0, tolerance_type="Dalton"),
                          sparse=True)

scores.stack_scores("retention_match", MetadataMatch(field="retention_time", tolerance=60.0), 
                    selector=("precursor_match"))
scores.stack_scores("cosine_score", CosineGreedy(), 
                    selector=("retention_match"))
scores.stack_scores("spec2vec", Spec2Vec(), 
                    selector=("cosine_score", 0.5))
```
And, as Niek suggested, the selector could also allow operations such as `("cosine_score", highest_per_reference=20)`

It would also be possible to add something like the scikit-learn `Pipeline` class:
```python
steps=[("precursor_match", PrecursorMzMatch(tolerance=50.0, tolerance_type="Dalton")),
        ("retention_match", MetadataMatch(field="retention_time", tolerance=60.0),
         selector=("precursor_match")),
        ("cosine_score", CosineGreedy(),
         selector=("retention_match"))]
scores = scores_pipeline(references=spectrums_1,
                          queries=spectrums_2, steps)
```

**Comment by hechth** [comment]
@florian-huber With the new Metadata class that will come soon, I think a first step would be to implement the `MetadataMatch` masking system, maybe even before formalizing the masking and pipeline system, as this single component I think will be needed anyway, it will just move from matching to masking.

I thought that the `MetadataMatch` could also leverage the file with defined conversions to only allow matches on _harmonized keys_, to have some consistency - how do you see this?

**Comment by florian-huber** [comment]
Hi @hechth , I started sketching `MetadataMatch` in #315 . Feel free to add comments/suggestions/issues!

**Comment by hechth** [comment]
@florian-huber I also took some time today to think about the scores pipelines and the individual classes and made some notes regarding the individual classes and what they have to be able to do in order to have a comprehensive and well-performing pipeline.

![image](https://user-images.githubusercontent.com/12066490/153248854-f8feb5c1-21ce-43ea-8b5c-06a9d556c64e.png)

My suggestion would be to continue with the following steps

- [x] Make `Scores` class store scores as sparse matrix -> this will only be the output, so performance is not crucial while if we want sparse computation, the result will be sparse anyway
- [x] Implement `sparse_matrix` functions for similarity metrics
- [x] Implement `Pipeline` class which takes the computation style as parameter and multiplies the result matrices from the steps with each other and returns a `Scores` object

If the `BaseSimilarity` class returns a different representation of the results based on the method, this is okay and can be harmonized when passing it to the scores class, so that `calculate_scores` can be used as before and we don't break the API.
When used in a `Pipeline`, the computation mode (pair, matrix or sparse) has to be defined and then the pipeline will simply compute all steps independently and multiply the outputs. This can be also speeded up via _task parallelism_.

**Comment by florian-huber** [comment]
Thanks for the suggestions and thoughts!
That seems very close to what I also had in mind.

One thing I actually first tried, but then wasn't fully happy with is the CSR sparse matrix (e.g. from scipy). I hence started building a custom sparse array class which I now sketch in #316.
The idea of this is that multiple scores can be added on top of each other and that we can apply a sort of "funneling" to the entire stack. What I mean by this is the following:

```python
import numpy as np
from matchms import StackedSparseScores

scores1 = np.random.random((12, 10))  # that would be a first score, maybe something cheap to calculate

stacked_array = StackedSparseScores(12, 10)
stacked_array.add_dense_matrix(scores1, "scores_1")
# Now we can remove all entries which are not within a desired range
stacked_array.filter_by_range("scores_1", low=0.5)  # not sure "filter" or "mask" ...? 

# We can then add more scores
scores2 = np.random.random((12, 10))
stacked_array.add_dense_matrix(scores2, "scores_2")  # will only add those scores for which "scores_1" exists
stacked_array.filter_by_range("scores_2", low=0.1, high=0.4)  # this will remove all scores for which "scores_2" is out of the range

# At the end we can access individual values, rows, columns
print(stacked_array[:, -1])
# Or we export all entires of a certain score to a dense (numpy) or sparse (scipy) array.
scores2_after_filtering = matrix.toarray("scores_2")
```

In addition to this, I agree that we should build a new **pipeline class** that will handle the different steps and interact with the sparse array.
For a start, I now added a new `.sparse_array()` method in the `BaseSimilarity` class. The pipeline could then take the current indices of a sparse array (`stacked_array.row` and `stacked_array.col`) and have the similarity score of choice compute the respective values, add them to a new layer of the stacked sparse array and (if desired) apply furhter filtering/masking.

**Comment by hechth** [comment]
What is the issue with the scipy sparse matrix? I think the overhead of coding a sparse matrix yourself shouldn't be underestimated.

**Comment by florian-huber** [comment]
There's a few things I wanted the sparse matrix to do well, mostly:
- add individual elements, say `matrix[17, 2501] = 0.7` when we compute new scores
- slicing, at least to some extend so that we can for instance quickly get the highest scores of a row or column

CSR (or CSC) sparse matrices from scipy are good for slicing, but adding individual values very expensive. COO matrices on the other hand do not offer slicing (at least not out of the box). That was a reason for me to look further.
In addition I had a pipeline in mind we worked on recently to do large-scale library matching. There, we would add many different scores and are only interested in the results that are within a range for all those computed scores. In that case I thought it would be neat to have a stacked array in COO style because that would give a lot of extra compression (the col/row indices could be the same for all scores which removes 2/3 of the stored values for the following scores).

Anyway, while my first tests looked very promising in terms of performance, the version I built now which includes more slicing options got much slower.
With many sparse types that was either not both possible (and hence requires converting one type into another), or it was very slow.

**Comment by florian-huber** [comment]
We could first start with CSR matrices. But that will require quiet some changes. Currently, the scores we have can make use of datatypes such as `[("score", numpy.float64), ("matches", "int")]` which allowes to combine different aspects of a score in one matrix. I don't think this is supported by scipy.

**Comment by florian-huber** [comment]
I implemented a first functional version in #316 now.
It does break the current API in some cases (that's why the docstring tests still fail, but all unit tests pass).

The idea here would be that
- `StackedSparseScores` is a sparse 2,5D array which is only meant for internal use. It will deliver a high compression and all methods needed for `Scores` or the new pipeline class.
- `Scores` will define the API to interact with all scores. Currently this is by running `scores.scores_by_query`, `scores.scores_by_reference`, or via `scores.scores`. A new alternative is `scores.get_scores_array()` which can return a numpy or a scipy-coo array. 
- `Scores` (and/or a new pipeline class) will have to decide which similarity calculation method to run. I imagine that for a dense (or nearly dense) case it would call `similarity_score.matrix()` and for sparse cases `similarity_score.sparse_array()`.
- The `similarity_score.sparse_array()` method is also where (in the future) we should/could add parallel execution.


Some quick first check:
```python
# do some benchmarking
import numpy as np
import time
from matchms import Spectrum, calculate_scores
from matchms.similarity import CosineGreedy, PrecursorMzMatch

np.random.seed(0)
spectrums = []
for i in range(1000):
    mz = np.random.randint(0, 1000, size=(10)).astype("float")
    mz.sort()
    spectrum = Spectrum(mz=mz,
                        intensities=np.random.random(size=(10)),
                        metadata = {"precursor_mz": 100.0 + np.random.randint(0, 500)})
    spectrums.append(spectrum)


# Version 1, compute scores for all spectrum pairs
tstart = time.time()
scores1 = calculate_scores(spectrums, spectrums, PrecursorMzMatch(tolerance=5.0))
scores2 = calculate_scores(spectrums, spectrums, CosineGreedy())
print(time.time() - tstart)  # => about 33s


# Version 2, compute sparse scores in 2nd step (only for entries which have precursor_mz within 5.0 Da)
cosine_greedy = CosineGreedy()

tstart = time.time()
scores = calculate_scores(spectrums, spectrums, PrecursorMzMatch(tolerance=5.0))
new_scores = cosine_greedy.sparse_array(spectrums, spectrums,
                                        scores.scores.row, scores.scores.col)
scores._scores.add_sparse_data(new_scores, "cosine_scores")   # not the intended API yet (->pipeline)
print(time.time() - tstart)  # => about 0.7s
```
It's of course an artificial example, but the gain in performance should even be larger when running on larger data (and spectra with more peaks).

**Comment by florian-huber** [comment]
@hechth how would you see the `Pipeline` class?

- Should it resemble a [scikit-learn style](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html) ?
- Or rather in the direction of [snakemake](https://f1000research.com/articles/10-33/v2) ?
- Or something else entirely?

And what should it cover?
- importing?
- filtering?
- score computations and masking

**Comment by hechth** [comment]
@florian-huber I was thinking about the Pipeline class as taking the query and reference data, the Similarity metric objects as steps and the mode of computation (pair/matrix/sparse) and return the resulting score object. This resulting score object could either have the individual scores or the composite score, or both.

I think that would go more into the direction of scikit learn than snakemake.

But by now I'm not thinking whether a `compose_scores` function which composes multiple score objects wouldn't actually make more sense? Also, since the sparse multiplications are not that straightforward/more intricate with little performance bonus, maybe different strategies to improve the computation could be employed.

Edit: This would maybe also be a moment to address #210  and #227

**Comment by florian-huber** [comment]
This is outdated and a `Pipeline` class as well as a sparse implementation in `Scores` has been added in #370.
(some of the discussions in here might come back in form of new issues though)
